File: src/main/scala/g0301_0400/s0394_decode_string/Solution.scala
package g0301_0400.s0394_decode_string

// #Medium #Top_100_Liked_Questions #String #Stack #Recursion #Level_1_Day_14_Stack #Udemy_Strings
// #Big_O_Time_O(n)_Space_O(n) #2023_11_09_Time_449_ms_(50.00%)_Space_56.5_MB_(45.45%)

object Solution {
    def decodeString(s: String): String = {
        val brStack = scala.collection.mutable.Stack[Int]()
        val stStack = scala.collection.mutable.Stack[String]()
        stStack.push("")
        var i = 0
        while (i < s.length) {
            if (s(i) - '0' <= 9) {
                var j = ""
                while (s(i) != '[') {
                    j += s(i)
                    i += 1
                }
                brStack.push(j.toInt)
                stStack.push("")
            } else if (s(i) == ']') {
                val o = brStack.pop()
                val o1 = stStack.pop()
                var r = ""
                for (_ <- 0 until o) r += o1
                stStack.push(stStack.pop() + r)
            } else {
                stStack.push(stStack.pop() + s(i))
            }
            i += 1
        }
        stStack.pop()
    }
}

File: src/main/rust/g0301_0400/s0394_decode_string/Solution.rs
// #Medium #Top_100_Liked_Questions #String #Stack #Recursion #Level_1_Day_14_Stack #Udemy_Strings
// #Big_O_Time_O(n)_Space_O(n) #2024_09_12_Time_0_ms_(100.00%)_Space_2.1_MB_(24.04%)

impl Solution {
    pub fn decode_string(s: String) -> String {
        s
        .chars()
        .fold((String::new(), String::new(), Vec::new()), |(mut next_string, mut next_integer, mut stack): (String, String, Vec<(usize, String)>), c| {
            match c {
                '0'..='9' => {
                    next_integer.push(c);
                    (next_string, next_integer, stack)
                },
                '[' => {
                    stack.push((next_integer.parse::<usize>().unwrap(), next_string));
                    next_integer.clear();
                    (String::new(), next_integer, stack)
                },
                ']' => {
                    let (last_integer, left_string) = stack.pop().unwrap();
                    let next_string = left_string + &next_string.repeat(last_integer);
                    (next_string, next_integer, stack)
                },
                c => {
                    next_string.push(c);
                    (next_string, next_integer, stack)
                }
            }
        }).0
    }
}
